macros/one_hot_encoder.sql

```diff
@@ -1,56 +1,89 @@
-    select distinct {{ source_column }} from {{ source_table }}
-    order by 1
-    {% endset %}
-    {% set results = run_query(category_values_query) %}
-    {% if execute %}
-        {# Return the first column #}
-        {% set category_values = results.columns[0].values() %}
-    {% else %}
-        {% set category_values = [] %}
-    {% endif %}
-{% elif categories is not iterable or categories is string or categories is mapping %}
-    {% set error_message %}
-The `categories` parameter must contain a list of category values.
-    {% endset %}
-    {%- do exceptions.raise_compiler_error(error_message) -%}
-{%- else -%}
-    {% set category_values = categories %}
-{%- endif -%}
-{%- if handle_unknown!='ignore' -%}
-    {% set error_message %}
-The `one_hot_encoder` macro only supports an 'handle_unknown' value of 'ignore' at this time.
-    {% endset %}
-    {%- do exceptions.raise_compiler_error(error_message) -%}
-{%- endif -%}
-{{ adapter.dispatch('one_hot_encoder',packages=['dbt_ml_preprocessing'])(source_table,source_column,category_values,handle_unknown,include_columns) }}
-{%- endmacro %}
+{% macro one_hot_encoder(source_table, source_column, categories='auto', handle_unknown='ignore',include_columns='*', exclude_columns=none) %}
 
-{% macro snowflake__one_hot_encoder(source_table,source_column,category_values,handle_unknown,include_columns) %}
-select 
-{% for column in include_columns %}
-{{ source_table }}.{{ column }},
-{% endfor %}
-{% for category in category_values %}
-iff({{source_column}}='{{category}}',true,false) as {{source_column}}_{{category}}
-{% if not loop.last %}, {% endif %}
-{% endfor %}
-from {{ source_table }}
+    {%- if categories=='auto' -%}
+        {% set category_values_query %}
+            select distinct
+                {{ source_column }}
+            from
+                {{ source_table }}
+            order by 1
+        {% endset %}
+        {% set results = run_query(category_values_query) %}
+        {% if execute %}
+            {# Return the first column #}
+            {% set category_values = results.columns[0].values() %}
+        {% else %}
+            {% set category_values = [] %}
+        {% endif %}
+    {% elif categories is not iterable or categories is string or categories is mapping %}
+        {% set error_message %}
+    The `categories` parameter must contain a list of category values.
+        {% endset %}
+        {%- do exceptions.raise_compiler_error(error_message) -%}
+    {%- else -%}
+        {% set category_values = categories %}
+    {%- endif -%}
+    {%- if handle_unknown!='ignore' -%}
+        {% set error_message %}
+    The `one_hot_encoder` macro only supports an 'handle_unknown' value of 'ignore' at this time.
+        {% endset %}
+        {%- do exceptions.raise_compiler_error(error_message) -%}
+    {%- endif -%}
+    {%- if include_columns!='*' or exclude_columns is not none -%}
+        {%- if include_columns!='*' and exclude_columns is not none -%}
```

Comment:
the `if`,`elif`,`elif` flow here is a bit repititive/feels a bit janky here...I couldn't work out a way to shorten it..but surely there is a nicer way?

Reply:
I've tidied these up